Source/NSOrderedSet.m

```diff
@@ -0,0 +1,1894 @@
+			numIndices: (NSUInteger)count
+{
+  if (count > 0)
+    {
+      NSUInteger	to = 0;
+      NSUInteger	from = 0;
+      NSUInteger	i;
+      GS_BEGINITEMBUF(sorted, count, NSUInteger);
+
+      while (from < count)
+	{
+	  NSUInteger	val = indices[from++];
+
+	  i = to;
+	  while (i > 0 && sorted[i-1] > val)
+	    {
+	      i--;
+	    }
+	  if (i == to)
+	    {
+	      sorted[to++] = val;
+	    }
+	  else if (sorted[i] != val)
+	    {
+	      NSUInteger	j = to++;
+
+	      if (sorted[i] < val)
+		{
+		  i++;
+		}
+	      while (j > i)
+		{
+		  sorted[j] = sorted[j-1];
+		  j--;
+		}
+	      sorted[i] = val;
+	    }
+	}
+
+      if (to > 0)
+	{
+	  IMP	rem = [self methodForSelector: remSel];
+
+	  while (to--)
+	    {
+	      (*rem)(self, remSel, sorted[to]);
+	    }
+	}
+      GS_ENDITEMBUF();
+    }
+}
+
+- (void)removeObjectsAtIndexes:(NSIndexSet *)indexes
+{
+  NSUInteger count = [indexes count];
+  NSUInteger indexArray[count];
+
+  [indexes getIndexes: indexArray
+             maxCount: count
+         inIndexRange: NULL];
+
+  [self _removeObjectsFromIndices: indexArray
+		       numIndices: count];
+}
+
+- (void)removeObjectsInArray:(NSArray *)otherArray
+{
+  NSUInteger	c = [otherArray count];
+
+  if (c > 0)
+    {
+      NSUInteger	i;
+      IMP	get = [otherArray methodForSelector: oaiSel];
+      IMP	rem = [self methodForSelector: @selector(removeObject:)];
+
+      for (i = 0; i < c; i++)
+	(*rem)(self, @selector(removeObject:), (*get)(otherArray, oaiSel, i));
+    }
+}
+
+- (void)removeObjectsInRange:(NSRange)aRange
+{
+  NSUInteger	i;
+  NSUInteger	s = aRange.location;
+  NSUInteger	c = [self count];
+
+  i = aRange.location + aRange.length;
+
+  if (c < i)
+    i = c;
+
+  if (i > s)
+    {
+      IMP	rem = [self methodForSelector: remSel];
+
+      while (i-- > s)
+	{
+	  (*rem)(self, remSel, i);
+	}
+    }
+}
+
+- (void)removeAllObjects
+{
+  NSUInteger	c = [self count];
+
+  if (c > 0)
+    {
+      while (c--)
+	{
+	  [self removeObjectAtIndex: 0];
+	}
+    }  
+}
+
+- (void)replaceObjectAtIndex:(NSUInteger)index // required override...
+                  withObject:(id)object
+{
+  [self subclassResponsibility: _cmd];
+}
+
+- (void) replaceObjectsAtIndexes: (NSIndexSet *)indexes
+                     withObjects: (NSArray *)objects
+{
+  NSUInteger count = [indexes count], i = 0;
+  NSUInteger indexArray[count];
+  NSUInteger c = [self count];
+  NSRange range = NSMakeRange(0,c);
+  
+  GS_BEGINIDBUF(objs, count);
+
+  // Get the indexes
+  [indexes getIndexes: indexArray
+	     maxCount: count
+	 inIndexRange: NULL];
+
+  // Get the objects from this set
+  [self getObjects: objs range: range];
+
+  // Iterate over the indexes and replace the objs
+  for (i = 0; i < count; i++)
+    {
+      id obj = [objects objectAtIndex: i];
+      NSUInteger indx = indexArray[i];
+      objs[indx] = obj;
+    }
+
+  // Remove all objects in this set...
+  [self removeAllObjects];
```

Comment:
We should try to avoid removeAllObjects, as this will be very slow, when only a few objects change. Why not call removeObjectsAtIndexes and then loop over the indexes and insert the new objects one after another?
Or use the method replaceObjectAtIndex: withObject:

Reply:
Initially I did it this way to ensure that the indexes didn't change.

Reply:
Corrected

Reply:
You changed the other place where I made this comment, but not this one.